File: firebase/functions/lib/utils/template_utils.dart
import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:data_models/templates/template.dart';

import 'infra/firestore_utils.dart';

class TemplateUtils {
  static Template templateFromSnapshot(DocumentSnapshot snapshot) {
    if (snapshot.documentID == "misc") {
      return Template(
        id: 'misc',
        creatorId: '',
        collectionPath: snapshot.reference.path,
      );
    } else {
      return Template.fromJson(
        firestoreUtils.fromFirestoreJson(snapshot.data.toMap()),
      );
    }
  }
}

File: firebase/functions/src/utils/template_utils.ts
import * as admin from 'firebase-admin'
import { Template } from '../types'

export class TemplateUtils {
    static templateFromSnapshot(snapshot: admin.firestore.DocumentSnapshot): Template {
        if (snapshot.id === 'misc') {
            return {
                id: 'misc',
                creatorId: '',
                collectionPath: snapshot.ref.path,
            }
        }
        const data = snapshot.data() ?? {}
        return {
            ...(data as Partial<Template>),
            id: snapshot.id,
            collectionPath: snapshot.ref.parent.path,
        } as Template
    }
}
